data/transactions/application.go

```diff
@@ -170,12 +184,170 @@ type ApplicationCallTxnFields struct {
 	// method below!
 }
 
-// BoxRef names a box by the slot
+// ResourceRef names a single resource
+type ResourceRef struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	// Only one of these may be set
+	Address basics.Address    `codec:"d"`
+	Asset   basics.AssetIndex `codec:"s"`
+	App     basics.AppIndex   `codec:"p"`
+	Holding HoldingRef        `codec:"h"`
+	Locals  LocalsRef         `codec:"l"`
+	Box     BoxRef            `codec:"b"`
+}
+
+// Empty ResourceRefs are allowed, as they ask for a box quota bump.
+func (rr ResourceRef) Empty() bool {
+	return rr.Address.IsZero() && rr.Asset == 0 && rr.App == 0 &&
+		rr.Holding.Empty() && rr.Locals.Empty() && rr.Box.Empty()
+}
+
+// wellFormed checks that a ResourceRef is either empty or of only one kind, and
```

Comment:
probably requires some clarification that `rr` is one of `access` elements. Maybe a better interface would be a free function taking `access` and looping over it?

Reply:
I added some clarification.

The main wellFormed() function for app calls is pretty much what you're describing (though it has other checks too.  

It does this:
```
	for _, rr := range ac.Access {
		if err := rr.wellFormed(ac.Access, proto); err != nil {
			return err
		}
	}
```

Would you like it better if the whole loop was pulled into it's own function, instead of the "inner" part of the loop as `rr.wellFormed()`?